cuqi/experimental/mcmc/_cwmh.py

```diff
@@ -0,0 +1,207 @@
+        pass # All targets are valid
+
+    @ProposalBasedSamplerNew.proposal.setter 
+    def proposal(self, value):
+        fail_msg = "Proposal should be either None, "+\
+            f"{cuqi.distribution.Distribution.__class__.__name__} "+\
+            "conditioned only on 'location' and 'scale', lambda function, "+\
+            f"or {cuqi.distribution.Normal.__class__.__name__} conditioned "+\
+            "only on 'mean' and 'std'"
+
+        if value is None:
+            self._proposal = cuqi.distribution.Normal(
+                mean=lambda location: location,
+                std=lambda scale: scale,
+                geometry=self.dim,
+            )
+
+        elif isinstance(value, cuqi.distribution.Distribution) and sorted(
+            value.get_conditioning_variables()
+        ) == ["location", "scale"]:
+            self._proposal = value
+
+        elif isinstance(value, cuqi.distribution.Normal) and sorted(
+            value.get_conditioning_variables()
+        ) == ["mean", "std"]:
+            self._proposal = value(
+                mean=lambda location: location, std=lambda scale: scale
+            )
+
+        elif not isinstance(value, cuqi.distribution.Distribution) and callable(
+            value):
+            self._proposal = value
+
+        else:
+            raise ValueError(fail_msg)
+
+    def step(self):
+        """ Perform one step of the sampler by transitioning the current point
+        to a new point according to the sampler's transition kernel. """
+        # Propose state x_i_star used to update
+        # each component of x_t step by step
+        x_t = self.current_point.copy()
+        x_star = self.current_point.copy()
+        target_eval_t = self.current_target
+        if isinstance(self.proposal,cuqi.distribution.Distribution):
+            x_i_star = self.proposal(
+                location= self.current_point, scale=self.scale).sample()
+        else:
+            x_i_star = self.proposal(self.current_point, self.scale) 
+        acc = np.zeros(self.dim)
+
+        for j in range(self.dim):
+            # propose state
+            x_star[j] = x_i_star[j]
+
+            # evaluate target
+            target_eval_star = self.target.logd(x_star)
+
+            # ratio and acceptance probability
+            ratio = target_eval_star - target_eval_t  # proposal is symmetric
+            alpha = min(0, ratio)
+
+            # accept/reject
+            u_theta = np.log(np.random.rand())
+            if (u_theta <= alpha):
+                x_t[j] = x_i_star[j]
+                target_eval_t = target_eval_star
+                acc[j] = 1
+            else:
+                pass
+                # x_t[j]       = x_t[j]
+                # target_eval_t = target_eval_t
+            x_star = x_t.copy()
+
+        self.current_target = target_eval_t
+        self.current_point = x_t
+        # NEW: update return
+        # return x_t, target_eval_t, acc
+        return acc
```

Comment:
In a way I like that you copied the sampler "step" explicitly. But since we are comparing with regression tests, I think it would be worthwhile to "clean up" some of the comments and improve the overall code quality of the step method in this new module.

Same comment holds for the warmup method.

Reply:
Nice, I updated the code for better readability, style and documentation

Reply:
Thanks! It looks much better now.